File: clients/android/kit/src/main/kotlin/io/unom/punktfunk/kit/Keymap.kt
package io.unom.punktfunk.kit

import android.view.KeyEvent

/**
 * Android `KEYCODE_*` → Windows Virtual-Key code (the punktfunk wire contract; the host maps VK →
 * evdev via `inject::vk_to_evdev`). The Android analogue of the Linux client's evdev→VK table
 * (`punktfunk-client-linux/src/keymap.rs`) and the Apple client's `hidToVK`. Positional/US-layout —
 * we forward the physical key, not the typed character. Unmapped keys → 0 (the Rust side drops them).
 * Extend this alongside `punktfunk-host/src/inject.rs::vk_to_evdev` (emit only VKs the host knows).
 */
object Keymap {
    fun toVk(keyCode: Int): Int = when (keyCode) {
        in KeyEvent.KEYCODE_A..KeyEvent.KEYCODE_Z -> 0x41 + (keyCode - KeyEvent.KEYCODE_A) // A–Z
        in KeyEvent.KEYCODE_0..KeyEvent.KEYCODE_9 -> 0x30 + (keyCode - KeyEvent.KEYCODE_0) // 0–9 row
        in KeyEvent.KEYCODE_F1..KeyEvent.KEYCODE_F12 -> 0x70 + (keyCode - KeyEvent.KEYCODE_F1) // F1–F12
        in KeyEvent.KEYCODE_NUMPAD_0..KeyEvent.KEYCODE_NUMPAD_9 ->
            0x60 + (keyCode - KeyEvent.KEYCODE_NUMPAD_0) // numpad 0–9

        // Whitespace / editing
        KeyEvent.KEYCODE_DEL -> 0x08 // Backspace (Android KEYCODE_DEL == backspace)
        KeyEvent.KEYCODE_TAB -> 0x09
        KeyEvent.KEYCODE_ENTER, KeyEvent.KEYCODE_NUMPAD_ENTER -> 0x0D
        KeyEvent.KEYCODE_ESCAPE -> 0x1B
        KeyEvent.KEYCODE_SPACE -> 0x20
        KeyEvent.KEYCODE_CAPS_LOCK -> 0x14
        KeyEvent.KEYCODE_BREAK -> 0x13 // Pause
        KeyEvent.KEYCODE_SYSRQ -> 0x2C // PrintScreen
        KeyEvent.KEYCODE_INSERT -> 0x2D
        KeyEvent.KEYCODE_FORWARD_DEL -> 0x2E // Delete (forward)
        KeyEvent.KEYCODE_NUM_LOCK -> 0x90
        KeyEvent.KEYCODE_SCROLL_LOCK -> 0x91

        // Navigation
        KeyEvent.KEYCODE_PAGE_UP -> 0x21
        KeyEvent.KEYCODE_PAGE_DOWN -> 0x22
        KeyEvent.KEYCODE_MOVE_END -> 0x23
        KeyEvent.KEYCODE_MOVE_HOME -> 0x24
        KeyEvent.KEYCODE_DPAD_LEFT -> 0x25
        KeyEvent.KEYCODE_DPAD_UP -> 0x26
        KeyEvent.KEYCODE_DPAD_RIGHT -> 0x27
        KeyEvent.KEYCODE_DPAD_DOWN -> 0x28

        // Modifiers (L/R-specific VKs; the host folds the generic ones onto the left variant)
        KeyEvent.KEYCODE_SHIFT_LEFT -> 0xA0
        KeyEvent.KEYCODE_SHIFT_RIGHT -> 0xA1
        KeyEvent.KEYCODE_CTRL_LEFT -> 0xA2
        KeyEvent.KEYCODE_CTRL_RIGHT -> 0xA3
        KeyEvent.KEYCODE_ALT_LEFT -> 0xA4
        KeyEvent.KEYCODE_ALT_RIGHT -> 0xA5 // AltGr
        KeyEvent.KEYCODE_META_LEFT -> 0x5B // Super/Win
        KeyEvent.KEYCODE_META_RIGHT -> 0x5C
        KeyEvent.KEYCODE_MENU -> 0x5D // Application

        // Numpad operators
        KeyEvent.KEYCODE_NUMPAD_MULTIPLY -> 0x6A
        KeyEvent.KEYCODE_NUMPAD_ADD -> 0x6B
        KeyEvent.KEYCODE_NUMPAD_SUBTRACT -> 0x6D
        KeyEvent.KEYCODE_NUMPAD_DOT -> 0x6E
        KeyEvent.KEYCODE_NUMPAD_DIVIDE -> 0x6F

        // OEM punctuation (US-layout positional)
        KeyEvent.KEYCODE_SEMICOLON -> 0xBA
        KeyEvent.KEYCODE_EQUALS -> 0xBB
        KeyEvent.KEYCODE_COMMA -> 0xBC
        KeyEvent.KEYCODE_MINUS -> 0xBD
        KeyEvent.KEYCODE_PERIOD -> 0xBE
        KeyEvent.KEYCODE_SLASH -> 0xBF
        KeyEvent.KEYCODE_GRAVE -> 0xC0
        KeyEvent.KEYCODE_LEFT_BRACKET -> 0xDB
        KeyEvent.KEYCODE_BACKSLASH -> 0xDC
        KeyEvent.KEYCODE_RIGHT_BRACKET -> 0xDD
        KeyEvent.KEYCODE_APOSTROPHE -> 0xDE

        else -> 0 // unmapped → Rust drops it
    }
}

File: crates/punktfunk-client-windows/src/keymap.rs
//! Local key/button codes → the punktfunk input wire contract.
//!
//! The wire carries Windows Virtual-Key codes (the GameStream convention; the host maps them
//! back with `inject::vk_to_evdev`). On Windows the VK is the *native* source — but winit
//! hands us a layout-independent **physical** `KeyCode` (`KeyA` is always the QWERTY-A
//! position), which is exactly what a game wants: positional keys map positionally regardless
//! of the user's keyboard layout. This table is that physical-position → VK mapping, the
//! direct analogue of the Linux client's evdev table.

use winit::keyboard::KeyCode;

/// Map a winit physical `KeyCode` to the Windows VK code the host expects. `None` = a key the
/// wire contract doesn't cover (media keys etc.) — drop it rather than guess.
pub fn keycode_to_vk(code: KeyCode) -> Option<u8> {
    use KeyCode::*;
    Some(match code {
        // --- Navigation / editing / whitespace ---
        Backspace => 0x08,
        Tab => 0x09,
        Enter => 0x0D,
        Pause => 0x13,
        CapsLock => 0x14,
        Escape => 0x1B,
        Space => 0x20,
        PageUp => 0x21,
        PageDown => 0x22,
        End => 0x23,
        Home => 0x24,
        ArrowLeft => 0x25,
        ArrowUp => 0x26,
        ArrowRight => 0x27,
        ArrowDown => 0x28,
        PrintScreen => 0x2C,
        Insert => 0x2D,
        Delete => 0x2E,

        // --- Digit row ---
        Digit0 => 0x30,
        Digit1 => 0x31,
        Digit2 => 0x32,
        Digit3 => 0x33,
        Digit4 => 0x34,
        Digit5 => 0x35,
        Digit6 => 0x36,
        Digit7 => 0x37,
        Digit8 => 0x38,
        Digit9 => 0x39,

        // --- Letters ---
        KeyA => 0x41,
        KeyB => 0x42,
        KeyC => 0x43,
        KeyD => 0x44,
        KeyE => 0x45,
        KeyF => 0x46,
        KeyG => 0x47,
        KeyH => 0x48,
        KeyI => 0x49,
        KeyJ => 0x4A,
        KeyK => 0x4B,
        KeyL => 0x4C,
        KeyM => 0x4D,
        KeyN => 0x4E,
        KeyO => 0x4F,
        KeyP => 0x50,
        KeyQ => 0x51,
        KeyR => 0x52,
        KeyS => 0x53,
        KeyT => 0x54,
        KeyU => 0x55,
        KeyV => 0x56,
        KeyW => 0x57,
        KeyX => 0x58,
        KeyY => 0x59,
        KeyZ => 0x5A,

        // --- Meta / context-menu ---
        SuperLeft => 0x5B,  // VK_LWIN
        SuperRight => 0x5C, // VK_RWIN
        ContextMenu => 0x5D,

        // --- Numpad ---
        Numpad0 => 0x60,
        Numpad1 => 0x61,
        Numpad2 => 0x62,
        Numpad3 => 0x63,
        Numpad4 => 0x64,
        Numpad5 => 0x65,
        Numpad6 => 0x66,
        Numpad7 => 0x67,
        Numpad8 => 0x68,
        Numpad9 => 0x69,
        NumpadMultiply => 0x6A,
        NumpadAdd => 0x6B,
        NumpadEnter => 0x6C, // VK_SEPARATOR (matches the Linux client's KP_ENTER mapping)
        NumpadSubtract => 0x6D,
        NumpadDecimal => 0x6E,
        NumpadDivide => 0x6F,

        // --- Function keys ---
        F1 => 0x70,
        F2 => 0x71,
        F3 => 0x72,
        F4 => 0x73,
        F5 => 0x74,
        F6 => 0x75,
        F7 => 0x76,
        F8 => 0x77,
        F9 => 0x78,
        F10 => 0x79,
        F11 => 0x7A,
        F12 => 0x7B,

        // --- Locks ---
        NumLock => 0x90,
        ScrollLock => 0x91,

        // --- Left/right modifiers ---
        ShiftLeft => 0xA0,
        ShiftRight => 0xA1,
        ControlLeft => 0xA2,
        ControlRight => 0xA3,
        AltLeft => 0xA4,
        AltRight => 0xA5,

        // --- OEM punctuation (US-layout positions) ---
        Semicolon => 0xBA,     // VK_OEM_1
        Equal => 0xBB,         // VK_OEM_PLUS
        Comma => 0xBC,         // VK_OEM_COMMA
        Minus => 0xBD,         // VK_OEM_MINUS
        Period => 0xBE,        // VK_OEM_PERIOD
        Slash => 0xBF,         // VK_OEM_2
        Backquote => 0xC0,     // VK_OEM_3
        BracketLeft => 0xDB,   // VK_OEM_4
        Backslash => 0xDC,     // VK_OEM_5
        BracketRight => 0xDD,  // VK_OEM_6
        Quote => 0xDE,         // VK_OEM_7
        IntlBackslash => 0xE2, // VK_OEM_102 (the 102nd key)

        _ => return None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Spot-check positions against the Windows VK constants the host's `vk_to_evdev` knows.
    #[test]
    fn maps_known_positions() {
        assert_eq!(keycode_to_vk(KeyCode::KeyA), Some(0x41));
        assert_eq!(keycode_to_vk(KeyCode::KeyZ), Some(0x5A));
        assert_eq!(keycode_to_vk(KeyCode::Digit0), Some(0x30));
        assert_eq!(keycode_to_vk(KeyCode::Escape), Some(0x1B));
        assert_eq!(keycode_to_vk(KeyCode::F11), Some(0x7A));
        assert_eq!(keycode_to_vk(KeyCode::ShiftLeft), Some(0xA0));
        assert_eq!(keycode_to_vk(KeyCode::IntlBackslash), Some(0xE2));
        assert_eq!(keycode_to_vk(KeyCode::Numpad9), Some(0x69));
        // A key outside the wire contract is dropped, not guessed.
        assert_eq!(keycode_to_vk(KeyCode::AudioVolumeMute), None);
    }
}
